### ISSUE TITLE
Example code for cummulative normal distribution is wrong

### ISSUE THREAD

**Issue by rstub** [arrayfire/arrayfire#2078]
The example code for the cummulative normal distribution in
https://github.com/arrayfire/arrayfire/blob/master/examples/financial/black_scholes_options.cpp#L22
gives wrong results as can be seen by adapting the test code from http://www.johndcook.com/blog/cpp_phi/:
```
#include <iostream>
#include <math.h>
#include <arrayfire.h>

using namespace af;

// The following function is a modified version of http://www.johndcook.com/blog/cpp_phi/
// The example above references Handbook of Mathematical Functions by Abramowitz and Stegun

array cnd(array x)
{
    // constants
    const float a1 =  0.254829592;
    const float a2 = -0.284496736;
    const float a3 =  1.421413741;
    const float a4 = -1.453152027;
    const float a5 =  1.061405429;
    const float p  =  0.3275911;
    const float sqrt2 = sqrt(2.0);

    // Save the sign of x
    array xSign = sign(x);

    x = abs(x) / sqrt2;

    // A&S formula 7.1.26
    array t = 1.0f / (1.0f + p*x);
    array y = 1.0f + 0.5f * (((((a5*t + a4)*t) + a3)*t + a2)*t + a1)*t*exp(-x*x);

    return xSign * y + !xSign * (1 - y); // equivalent of (x >= 0) ? y : (1 - y);
}

void testCnd()
{
    // Select a few input values
    float x[] = {-3.0, -1.0, 0.0, 0.5, 2.1};
    array X(5, x, afHost);

    // Output computed by Mathematica
    // y = Phi[x]
    float y[] = {0.00134989803163, 0.158655253931, 0.5, 0.691462461274, 0.982135579437};
    array Y(5, y, afHost);
    
    float maxError = max<float>(abs(Y - cnd(X))); 
    std::cout << "Maximum error: " << maxError << "\n";
} 

int main() {
  testCnd();
  return 0;
}
```
This gives: `Maximum error: 1`

One could try to fix the above code, but since it uses one particular approximation to the error function `erf` which is already available, it is easier to just use [that relation](https://en.wikipedia.org/wiki/Error_function#Cumulative_distribution_function) directly:
```
array cnd(array x)
{
    const float sqrt05 = sqrt(0.5f);
    return 0.5f * erfc(- x * sqrt05);
}
```

If you want I can provide a PR for this.

**Comment by 9prady9** [comment]
@rstub So, the equations were wrong hence the maximum error of 1.

```diff
diff --git a/examples/financial/black_scholes_options.cpp b/examples/financial/black_scholes_options.cpp
index 5d979fd6..dbf81699 100644
--- a/examples/financial/black_scholes_options.cpp
+++ b/examples/financial/black_scholes_options.cpp
@@ -37,9 +37,9 @@ array cnd(array x)
 
     // A&S formula 7.1.26
     array t = 1.0f / (1.0f + p*x);
-    array y = 1.0f + 0.5f * (((((a5*t + a4)*t) + a3)*t + a2)*t + a1)*t*exp(-x*x);
+    array y = 1.0f - 0.5f * (((((a5*t + a4)*t) + a3)*t + a2)*t + a1)*t*exp(-x*x);
 
-    return xSign * y + !xSign * (1 - y); // equivalent of (x >= 0) ? y : (1 - y);
+    return !xSign * y + xSign * (1 - y); // equivalent of (x >= 0) ? y : (1 - y);
 }
 
 static void black_scholes(array& C, array& P,
```
The above diff gave an error of `Maximum error: 8.75443e-08`

If the erfc is functional equivalent of the above code, please do send us the PR. Thank you for reporting it. 

Note: Using `erfc` seems to have sped up the black scholes example by 2x, I am not sure if efrc does any approximations. @rstub If you can shed some light regarding that, I would appreciate it as I am not aware of this.

**Comment by rstub** [comment]
I have made a quick comparison in R (using RcppArrayFire) to compare four CDN implementations:

* pnorm is the CDN from R, which I am assuming as "exact"
* cdn1 is the corrected A&S formula 7.1.26
* cdn2 uses the relationship between CDN and erfc
* cdn3 uses the analogue relationship between CDN and erf

Calculating all these functions between -10 and 10 in steps of 0.1 on sees that for small values of x using erfc is able to capture the tails of the distribution, while the approximations based on erf (cdn1 and cdn2) are exactly zero:
```
x <- seq(from = -10, to = 10, by = 0.1)
cnd <- data.frame(
    x = x,
    pnorm = pnorm(x),
    cnd1 = cnd1(x),
    cnd2 = cnd2(x),
    cnd3 = cnd3(x)
)
head(cnd)
##       x        pnorm cnd1         cnd2 cnd3
## 1 -10.0 7.619853e-24    0 7.619854e-24    0
## 2  -9.9 2.081375e-23    0 2.081389e-23    0
## 3  -9.8 5.629282e-23    0 5.629280e-23    0
## 4  -9.7 1.507493e-22    0 1.507502e-22    0
## 5  -9.6 3.997221e-22    0 3.997217e-22    0
## 6  -9.5 1.049452e-21    0 1.049450e-21    0
```

The maximum difference between pnorm and cdn1 is 1.27e-07, while the maximum differences between pnorm and cdn2 or cdn3 are 3.19e-08 or 4.51e-08, respectivly. Comparing pnorm - cdn1 (dots) and pnorm - cdn2 (line) graphically:
![grafik](https://user-images.githubusercontent.com/27821131/37143728-49f1d594-22bc-11e8-8a69-4ae71deb3f4a.png)

As for performance, the erf and erfc based solutions are clearly faster than the A&S formula:
```
Unit: microseconds
  expr     min      lq      mean   median       uq     max neval
 pnorm  12.800  15.042  16.01323  16.1785  16.8845  24.684   100
  cnd1 187.558 194.095 221.29298 200.5095 208.2970 373.331   100
  cnd2 110.330 119.469 134.24501 122.0135 127.5200 699.267   100
  cnd3 110.674 119.032 130.28833 122.1515 125.5820 298.205   100
```

So for me cdn2 is the winner here. I will create an appropriate PR.
